packages/filesystem/src/browser-only/opfs-filesystem-provider.ts

```diff
@@ -21,36 +21,76 @@ import {
-import { Emitter, Event, URI, Disposable, Path } from '@theia/core';
+import { Emitter, Event, URI, Disposable, DisposableCollection, type CancellationToken } from '@theia/core';
+import { EncodingService } from '@theia/core/lib/common/encoding-service';
+import { BinaryBuffer } from '@theia/core/lib/common/buffer';
+import { TextDocumentContentChangeEvent } from '@theia/core/shared/vscode-languageserver-protocol';
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import { createWorker, type FileStat, type OPFSFileSystem } from 'opfs-worker';
 import { OPFSInitialization } from './opfs-filesystem-initialization';
-
-/** Options to be used when traversing the file system handles */
-interface CreateFileSystemHandleOptions {
-    isDirectory?: boolean;
-    create?: boolean;
-}
+import { ReadableStreamEvents, newWriteableStream } from '@theia/core/lib/common/stream';
+import { readFileIntoStream } from '../common/io';
 
 @injectable()
-export class OPFSFileSystemProvider implements FileSystemProviderWithFileReadWriteCapability {
-    capabilities: FileSystemProviderCapabilities = FileSystemProviderCapabilities.FileReadWrite;
+export class OPFSFileSystemProvider implements Disposable,
+    FileSystemProviderWithFileReadWriteCapability,
+    FileSystemProviderWithOpenReadWriteCloseCapability,
+    FileSystemProviderWithFileFolderCopyCapability {
+
+    private readonly BUFFER_SIZE = 64 * 1024;
+
+    capabilities: FileSystemProviderCapabilities =
+        FileSystemProviderCapabilities.FileReadWrite |
+        FileSystemProviderCapabilities.FileOpenReadWriteClose |
+        FileSystemProviderCapabilities.FileFolderCopy |
+        FileSystemProviderCapabilities.Update;
+
     onDidChangeCapabilities: Event<void> = Event.None;
 
     private readonly onDidChangeFileEmitter = new Emitter<readonly FileChange[]>();
+
     readonly onDidChangeFile = this.onDidChangeFileEmitter.event;
-    onFileWatchError: Event<void> = Event.None;
+    readonly onFileWatchError: Event<void> = Event.None;
 
     @inject(OPFSInitialization)
     protected readonly initialization: OPFSInitialization;
 
-    private directoryHandle: FileSystemDirectoryHandle;
+    @inject(EncodingService)
+    protected readonly encodingService: EncodingService;
+
+    private fs!: OPFSFileSystem;
+    private broadcastChannel!: BroadcastChannel;
     private initialized: Promise<true>;
 
+    protected readonly toDispose = new DisposableCollection(
+        this.onDidChangeFileEmitter
+    );
+
     @postConstruct()
     protected init(): void {
         const setup = async (): Promise<true> => {
-            this.directoryHandle = await this.initialization.getRootDirectory();
+            const rootHandler = await this.initialization.getRootDirectory();
+            // NOTE: FileSystemDirectoryHandle here for backward compatibility
+            const root = (typeof rootHandler === 'string') ? rootHandler : (rootHandler as FileSystemDirectoryHandle).name;
```

Comment:
I prefer remove this mess with FileSystemDirectoryHandle if its acceptable to break compability